### PR TITLE
Replace calling the gpg binary with the gpgme library

### PR DESCRIPTION
**Pull request by bzeller**

### CODE REVIEW

File: zypp/PublicKey_p.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+|                          / /__ | | | | | |                           |
+|                         /_____||_| |_| |_|                           |
+|                                                                      |
+\---------------------------------------------------------------------*/
+#include "zypp/PublicKey.h"
+
+namespace zypp {
+
+struct PublicSubkeyData::Impl
+{
```
Comment (mlandres) [inline zypp/PublicKey_p.h]: That's probably my fault. libzypp currently exports every symbol within the zypp namespace, that's why the Impl classes are kept inside the .cc file. I guess removing the `PublicKeyScanner` now requires to expose them.

I'd prefer to keep these classes inside the PublicKey.cc and introduce some (private?) setter in `PublicKeyData`.  Might be possible that `KeyManagerCtx` passes the `gpgme_key_t key` to the setter and the setter fills in the data?

File: libzypp.spec.cmake [hunk 2]
```diff
@@ -103,11 +103,11 @@ BuildRequires:  librpm-devel > 4.4
 %endif
 
 %if 0%{?suse_version}
-BuildRequires:  gpg2
-Requires:       gpg2
+BuildRequires:  libgpgme-devel
+Requires:       gpgme
```
[the remark this answers is not in the document]
Comment (bzeller) [reply libzypp.spec.cmake]: Right, that should be libgpgme11

File: libzypp.spec.cmake [hunk 3]
```diff
@@ -103,11 +103,11 @@ BuildRequires:  librpm-devel > 4.4
...
 %else
-BuildRequires:  gnupg2
-Requires:       gnupg2
+BuildRequires:  gpgme-devel
+Requires:       gpgme
```
[the remark this answers is not in the document]
Comment (bzeller) [reply libzypp.spec.cmake]: That was just a wild guess based on other similar entries. What would be the correct package here?

File: zypp/KeyManager.cc [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    return std::list<std::string>();
+
+  FILEPtr dataFile(fopen(file.c_str(), "rb"), fclose);
+  if(!dataFile)
+    return std::list<std::string>();
+
+  GpgmeDataPtr fileData(nullptr, gpgme_data_release);
+  gpgme_error_t err = gpgme_data_new_from_stream (&fileData.get(), dataFile.get());
+  if (err) {
+    ERR << "file" << __FILE__ << "line: "<<__LINE__
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.cc]: Sounds like a good idea

File: zypp/PublicKey_p.h (hunk 1 above)
Comment (bzeller) [reply zypp/PublicKey_p.h]: I thought about the same but came to the conclusion that it would be nice to not pollute the other parts of the source with gpgme related types. About the exporting, why not change to exporting symbols explicitely with the symbol visibility flag: https://gcc.gnu.org/wiki/Visibility

File: libzypp.spec.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlandres) [reply libzypp.spec.cmake]: No. RPM auto-requires linked libraries when packaging libs and binaries. You don't need to require them explicitly.

File: libzypp.spec.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (mlandres) [reply libzypp.spec.cmake]: Like the above. You don't need to require it explicitly.

File: zypp/PublicKey_p.h (hunk 1 above)
Comment (mlandres) [reply zypp/PublicKey_p.h]: Sooner or later we have to tackle the visibility issue. For now I'd feel more comfortable putting the private code back into the .cc file.

File: libzypp.spec.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (bzeller) [reply libzypp.spec.cmake]: done

File: libzypp.spec.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (bzeller) [reply libzypp.spec.cmake]: done

File: zypp/PublicKey_p.h (hunk 1 above)
Comment (bzeller) [reply zypp/PublicKey_p.h]: done

File: zypp/KeyManager.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/*---------------------------------------------------------------------\
+|                          ____ _   __ __ ___                          |
+|                         |__  / \ / / . \ . \                         |
+|                           / / \ V /|  _/  _/                         |
+|                          / /__ | | | | | |                           |
+|                         /_____||_| |_| |_|                           |
+|                                                                      |
+\---------------------------------------------------------------------*/
+/** \file	zypp/AutoDispose.h
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.h]: done

File: cmake/modules/FindGpgme.cmake [hunk 6]
```diff
@@ -0,0 +1,412 @@
+# Taken from gpgmepp project
```
[the remark this answers is not in the document]
Comment (bzeller) [reply cmake/modules/FindGpgme.cmake]: done

File: zypp/KeyManager.cc [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+  shared_ptr<KeyManagerCtx> me( new KeyManagerCtx());
+  me->_pimpl->_ctx = ctx;
+  return me;
+}
+
+/*!
+ * \brief KeyManagerCtx::setHomedir
+ * Changes the keyring directory
+ */
+bool KeyManagerCtx::setHomedir(const zypp::filesystem::Pathname &keyring_r)
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.cc]: done

File: zypp/KeyManager.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.cc]: done

File: zypp/KeyManager.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmacvicar) [reply zypp/KeyManager.h]: It would be good if you could encode these rules in something consumable by [clang-format](https://clang.llvm.org/docs/ClangFormat.html) , and then have travis run that on every PR/branch.

[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.h]: Yes, we already discussed using clang-format as git hook, so commited source code gets auto formatted. However we need to be careful so we do not make backporting too hard with all the whitespace changes. One solution would be to run clang-format on all our branches right away to keep the conflicts to a minimum.

File: zypp/base/String.h [hunk 8]
```diff
@@ -142,10 +142,10 @@ namespace zypp
 #endif
 
     inline std::string asString( const char * t )
-    { return t; }
+    { return t == nullptr ? std::string() : t; }
```
Comment (mlandres) [inline zypp/base/String.h]: Good catch. The changes here are not related to the issue. It's a separate 'bug' so it should go into a separate commit (and we will pick it for the maintenance branches as well).

File: zypp/KeyManager.cc [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+{
+  gpgme_check_version(NULL);
+
+}
+
+//output operator from gpgme_error_t
+std::ostream & operator<<( std::ostream & str, const gpgme_error_t err )
+{
+  return str << "<"<<gpgme_strsource(err)<<"> "<<gpgme_strerror(err);
+}
```
Comment (mlandres) [inline zypp/KeyManager.cc]: Did this work for you? `gpgme_error_t` is an integral type. My attempts to log a `gpgme_error_t err;` result in the plain number being written. A simple wrapper class could do the trick (similar to `zypp/base/Errno.h`):
```
namespace zypp
{
  struct GpgmeError
  {
    GpgmeError( gpgme_error_t err_r = GPG_ERR_NO_ERROR )
    : _err( err_r )
    {}
    operator gpgme_error_t() const { return _err; }
  private:
    const gpgme_error_t _err;
  };
  std::ostream & operator<<( std::ostream & str, const GpgmeError & obj )
  { return str << "<" << gpgme_strsource(obj) << "> " << gpgme_strerror(obj); }

  GpgmeError err;
  err = gpgme_op_...();
  if ( err )
      ERR << err << endl;
```

I was looking into this, because there are several locations, where errors occur, but are not logged. Not only gpgme errors, also e.g. a not existing signature file (`verifyAndReadSignaturesFprs`), etc.

In production environments we often can't ask the customer to try something different or to reproduce some error, so we should try to have all the info in the logfile we need to figure out what happened. E.g `verifyAndReadSignaturesFprs`: Whether `!PathInfo( signature ).isExist()` or `!dataFile` or... led to returning an empty list should be visible in the logfile. If we figure out we log too much we can still reduce it.

File: zypp/KeyManager.h [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+\---------------------------------------------------------------------*/
+/** \file	zypp/KeyManager.h
+ *
+*/
+#ifndef ZYPP_KEYMANAGER_H
+#define ZYPP_KEYMANAGER_H
+
+#include <zypp/base/PtrTypes.h>
+#include <zypp/Pathname.h>
+#include <zypp/PublicKey.h>
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.h]: Done

File: zypp/KeyManager.h [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+        /** Reads all fingerprints from the \a signature file , returns a list of all found fingerprints */
+        std::list<std::string> readSignatureFingerprints(const Pathname & signature);
+
+    private:
+      class Impl;
+
+      KeyManagerCtx();
+
+      RW_pointer<Impl> _pimpl; /// Pointer to implementation
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.h]: Done

File: zypp/KeyManager.h [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+{
+    public:
+        typedef zypp::shared_ptr<KeyManagerCtx> Ptr;
+
+        /** Creates a new KeyManagerCtx for PGP */
+        static Ptr createForOpenPGP();
+
+        /** Changes the keyring directory */
+        bool setHomedir (const zypp::Pathname & keyring_r);
+        zypp::Pathname homedir ()const;
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyManager.h]: Done

File: zypp/PublicKey.h [hunk 13]
```diff
@@ -76,6 +81,8 @@ namespace zypp
     /** Default constructed: empty data. */
     PublicSubkeyData();
 
+    PublicSubkeyData(const gpgme_subkey_t rawSubKeyData);
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/PublicKey.h]: Done

File: zypp/PublicKey.h [hunk 14]
```diff
@@ -25,6 +25,10 @@
 #include "zypp/Edition.h"
 #include "zypp/Date.h"
 
+typedef struct gpgme_data * gpgme_data_t;
+typedef struct _gpgme_key * gpgme_key_t;
+typedef struct _gpgme_subkey *gpgme_subkey_t;
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/PublicKey.h]: The typedef is not stricly required anyway, also using the const with the typedef'ed pointer is a bad idea because it introduces some const weirdness.  I think I just replace those with _gpgme_key * directly and forget about the typedef completely

File: zypp/PublicKey.h [hunk 15]
```diff
@@ -112,7 +119,7 @@ namespace zypp
   private:
     class Impl;
     RWCOW_pointer<Impl> _pimpl;
-    friend class PublicKeyScanner;
+    friend class KeyManagerCtx;
     friend std::ostream & dumpOn( std::ostream & str, const PublicKeyData & obj );
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/PublicKey.h]: Done

File: zypp/PublicKey.h [hunk 16]
```diff
@@ -214,7 +223,11 @@ namespace zypp
   private:
     class Impl;
     RWCOW_pointer<Impl> _pimpl;
-    friend class PublicKeyScanner;
+
+    PublicKeyData(shared_ptr<Impl> data);
+
+    friend class KeyManagerCtx;
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/PublicKey.h]: Done

File: zypp/KeyManager.cc (hunk 9 above)
Comment (bzeller) [reply zypp/KeyManager.cc]: Ok, I was not sure how verbose I should be in the code with error logging. I will go over the code again and put some more output in.

File: zypp/KeyRing.cc [hunk 17]
```diff
@@ -30,15 +30,13 @@
 #include "zypp/ExternalProgram.h"
 #include "zypp/TmpPath.h"
 #include "zypp/ZYppCallbacks.h"       // JobReport::instance
+#include "zypp/KeyManager.h"
 
 using std::endl;
 
 #undef  ZYPP_BASE_LOGGER_LOGGROUP
 #define ZYPP_BASE_LOGGER_LOGGROUP "zypp::KeyRing"
```
[the remark this answers is not in the document]
Comment (bzeller) [reply zypp/KeyRing.cc]: Done

File: zypp/KeyManager.cc (hunk 9 above)
Comment (bzeller) [reply zypp/KeyManager.cc]: Done

File: zypp/KeyManager.cc [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+typedef boost::interprocess::scoped_ptr<_gpgme_key, boost::function<void (gpgme_key_t)>>  GpgmeKeyPtr;
+typedef boost::interprocess::scoped_ptr<FILE, boost::function<int (FILE *)>> FILEPtr;
+
+struct GpgmeErr
+{
+  GpgmeErr( gpgme_error_t err_r = GPG_ERR_NO_ERROR )
+  : _err( err_r )
+  {}
+  operator gpgme_error_t() const { return _err; }
+  GpgmeErr &operator= (gpgme_error_t err_r) { _err = err_r; return *this;}
```
Comment (mlandres) [inline zypp/KeyManager.cc]: Out of curiosity, did you need to explicitly define the operator= ?

Comment (bzeller) [reply zypp/KeyManager.cc]: Yes, did not compile without it, when I want to assign a new error value to a existing instance the = operator is needed. I guess using GpgmeErr & as argument would've worked too.

Comment (mlandres) [reply zypp/KeyManager.cc]: Which g++ version are you using? 
Just asking because the CTOR allows implicit conversion from gpgme_error_t to GpgmeErr, so trivial assignment should be possible (g++ 4.8.5 and 7.3.0 compiled it without the operator=).
